In [1]:
#by convention, we use these shorter two-letter names
import pysal as ps
import pandas as pd
import numpy as np

# Spatial Data Processing with PySAL & Pandas

PySAL has two simple ways to read in data. But, first, you need to get the path from where your notebook is running on your computer to the place the data is. For example, to find where the notebook is running:

In [2]:
!pwd

/home/serge/Dropbox/p/pysal/workshops/scipy16/gds_scipy16/content/part1


PySAL has a command that it uses to get the paths of its example datasets. Let's work with a commonly-used dataset first. 

In [3]:

dbf_path = "../data/texas.dbf"
print(dbf_path)

../data/texas.dbf


For the purposes of this part of the workshop, we'll use the `texas.dbf` example data, and the `usjoin.csv` data. 

In [4]:
csv_path = ps.examples.get_path('usjoin.csv')

### Working with shapefiles

To read in a shapefile, we will need the path to the file.

In [5]:
shp_path = '../data/texas.shp'
print(shp_path)

../data/texas.shp


Then, we open the file using the `ps.open` command:

In [6]:
f = ps.open(shp_path)

`f` is what we call a "file handle." That means that it only *points* to the data and provides ways to work with it. By itself, it does not read the whole dataset into memory. To see basic information about the file, we can use a few different methods. 

For instance, the header of the file, which contains most of the metadata about the file:

In [7]:
f.header

{'BBOX Mmax': 0.0,
 'BBOX Mmin': 0.0,
 'BBOX Xmax': -93.50721740722656,
 'BBOX Xmin': -106.6495132446289,
 'BBOX Ymax': 36.49387741088867,
 'BBOX Ymin': 25.845197677612305,
 'BBOX Zmax': 0.0,
 'BBOX Zmin': 0.0,
 'File Code': 9994,
 'File Length': 49902,
 'Shape Type': 5,
 'Unused0': 0,
 'Unused1': 0,
 'Unused2': 0,
 'Unused3': 0,
 'Unused4': 0,
 'Version': 1000}

To actually read in the shapes from memory, you can use the following commands:

In [8]:
f.by_row(14) #gets the 14th shape from the file

In [9]:
all_polygons = f.read() #reads in all polygons from memory

In [10]:
len(all_polygons)

254

So, all 254 polygons have been read in from file. These are stored in PySAL shape objects, which can be used by PySAL and can be converted to other Python shape objects. ]

They typically have a few methods. So, since we've read in polygonal data, we can get some properties about the polygons. Let's just have a look at the first polygon:

In [11]:
all_polygons

In [12]:
all_polygons[0].centroid #the centroid of the first polygon

(-100.27156110567945, 36.27508640938005)

In [13]:
all_polygons[0].area

0.23682222998468205

In [14]:
all_polygons[0].perimeter

1.9582821721538344

While in the Jupyter Notebook, you can examine what properties an object has by using the tab key.

In [15]:
polygon = all_polygons[0]

In [16]:
polygon. #press tab when the cursor is right after the dot

SyntaxError: invalid syntax (<ipython-input-16-aa03438a2fa8>, line 1)

### Working with Data Tables

When you're working with tables of data, like a `csv` or `dbf`, you can extract your data in the following way. Let's open the dbf file we got the path for above.

In [17]:
f = ps.open(dbf_path)

Just like with the shapefile, we can examine the header of the dbf file

In [18]:
f.header

[u'NAME',
 u'STATE_NAME',
 u'STATE_FIPS',
 u'CNTY_FIPS',
 u'FIPS',
 u'STFIPS',
 u'COFIPS',
 u'FIPSNO',
 u'SOUTH',
 u'HR60',
 u'HR70',
 u'HR80',
 u'HR90',
 u'HC60',
 u'HC70',
 u'HC80',
 u'HC90',
 u'PO60',
 u'PO70',
 u'PO80',
 u'PO90',
 u'RD60',
 u'RD70',
 u'RD80',
 u'RD90',
 u'PS60',
 u'PS70',
 u'PS80',
 u'PS90',
 u'UE60',
 u'UE70',
 u'UE80',
 u'UE90',
 u'DV60',
 u'DV70',
 u'DV80',
 u'DV90',
 u'MA60',
 u'MA70',
 u'MA80',
 u'MA90',
 u'POL60',
 u'POL70',
 u'POL80',
 u'POL90',
 u'DNL60',
 u'DNL70',
 u'DNL80',
 u'DNL90',
 u'MFIL59',
 u'MFIL69',
 u'MFIL79',
 u'MFIL89',
 u'FP59',
 u'FP69',
 u'FP79',
 u'FP89',
 u'BLK60',
 u'BLK70',
 u'BLK80',
 u'BLK90',
 u'GI59',
 u'GI69',
 u'GI79',
 u'GI89',
 u'FH60',
 u'FH70',
 u'FH80',
 u'FH90']

So, the header is a list containing the names of all of the fields we can read. If we were interested in getting the `['NAME', 'STATE_NAME', 'HR90', 'HR80']` fields.

If we just wanted to grab the data of interest, `HR90`, we can use either `by_col` or `by_col_array`, depending on the format we want the resulting data in:

In [19]:
HR90 = f.by_col('HR90')
print(type(HR90).__name__, HR90[0:5])
HR90 = f.by_col_array('HR90')
print(type(HR90).__name__, HR90[0:5])

('list', [0.0, 0.0, 18.31166453, 0.0, 3.6517674554])
('ndarray', array([[  0.        ],
       [  0.        ],
       [ 18.31166453],
       [  0.        ],
       [  3.65176746]]))


As you can see, the `by_col` function returns a list of data, with no shape. It can only return one column at a time:

In [20]:
HRs = f.by_col('HR90', 'HR80')

TypeError: __call__() takes exactly 2 arguments (3 given)

This error message is called a "traceback," as you see in the top right, and it usually provides feedback on why the previous command did not execute correctly. Here, you see that one-too-many arguments was provided to `__call__`, which tells us we cannot pass as many arguments as we did to `by_col`.

If you want to read in many columns at once and store them to an array, use `by_col_array`:

In [21]:
HRs = f.by_col_array('HR90', 'HR80')

In [22]:
HRs

array([[  0.        ,   0.        ],
       [  0.        ,  10.50199538],
       [ 18.31166453,   5.10386362],
       [  0.        ,   0.        ],
       [  3.65176746,  10.4297038 ],
       [  0.        ,   0.        ],
       [  0.        ,  18.85369532],
       [  2.59514448,   6.33617194],
       [  0.        ,   0.        ],
       [  5.59753708,   6.0331825 ],
       [ 17.3692707 ,  14.53139626],
       [  5.06893755,   4.9960032 ],
       [  9.73560868,   8.84307335],
       [  5.66989851,   4.67049647],
       [  0.        ,   0.        ],
       [  4.08893052,   4.44077341],
       [ 16.49348507,   0.        ],
       [  0.        ,   7.17154332],
       [  0.        ,   8.1799591 ],
       [  3.48074279,   7.8746358 ],
       [  0.        ,   6.0397415 ],
       [  3.67511944,  12.63104711],
       [  8.19705726,  17.14148582],
       [  0.        ,   0.        ],
       [ 11.19883532,   4.79616307],
       [  8.53606487,   5.95876534],
       [  6.30954634,   5.23450586],
 

It is best to use `by_col_array` on data of a single type. That is, if you read in a lot of columns, some of them numbers and some of them strings, all columns will get converted to the same datatype:

In [23]:
allcolumns = f.by_col_array(['NAME', 'STATE_NAME', 'HR90', 'HR80'])

In [24]:
allcolumns

array([[u'Lipscomb', u'Texas', u'0.0', u'0.0'],
       [u'Sherman', u'Texas', u'0.0', u'10.501995379'],
       [u'Dallam', u'Texas', u'18.31166453', u'5.1038636248'],
       ..., 
       [u'Hidalgo', u'Texas', u'7.3003167816', u'8.2383277607'],
       [u'Willacy', u'Texas', u'5.6481219994', u'7.6212251119'],
       [u'Cameron', u'Texas', u'12.302014455', u'11.761321464']], 
      dtype='<U13')

Note that the numerical columns, `HR90` & `HR80` are now considered strings, since they show up with the single tickmarks around them, like `'0.0'`.

These methods work similarly for `.csv` files as well

### Using Pandas with PySAL

A new functionality added to PySAL recently allows you to work with shapefile/dbf pairs using Pandas. This *optional* extension is only turned on if you have Pandas installed. The extension is the `ps.pdio` module:

In [25]:
ps.pdio

<module 'pysal.contrib.pdutilities' from '/home/serge/anaconda2/envs/pydata/lib/python2.7/site-packages/pysal/contrib/pdutilities/__init__.pyc'>

To use it, you can read in shapefile/dbf pairs using the `ps.pdio.read_files` command. 

In [26]:
shp_path = ps.examples.get_path('NAT.shp')
data_table = ps.pdio.read_files(shp_path)

This reads in *the entire database table* and adds a column to the end, called `geometry`, that stores the geometries read in from the shapefile. 

Now, you can work with it like a standard pandas dataframe.

In [27]:
data_table.head()

NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
0  Lake of the Woods   Minnesota         27       077  27077      27      77   
1              Ferry  Washington         53       019  53019      53      19   
2            Stevens  Washington         53       065  53065      53      65   
3           Okanogan  Washington         53       047  53047      53      47   
4       Pend Oreille  Washington         53       051  53051      53      51   

   FIPSNO  SOUTH      HR60                        ...                          \
0   27077      0  0.000000                        ...                           
1   53019      0  0.000000                        ...                           
2   53065      0  1.863863                        ...                           
3   53047      0  2.612330                        ...                           
4   53051      0  0.000000                        ...                           

      BLK90      GI59      GI69      GI79      GI89       FH60  FH70  \
0  0.024534  0.285235  0.372336  0.342104  0.336455  11.279621   5.4   
1  0.317712  0.256158  0.360665  0.361928  0.360640  10.053476   2.6   
2  0.210030  0.283999  0.394083  0.357566  0.369942   9.258437   5.6   
3  0.155922  0.258540  0.371218  0.381240  0.394519   9.039900   8.1   
4  0.134605  0.243263  0.365614  0.358706  0.387848   8.243930   4.1   

        FH80       FH90                                           geometry  
0   5.663881   9.515860  <pysal.cg.shapes.Polygon object at 0x7f1e542b5...  
1  10.079576  11.397059  <pysal.cg.shapes.Polygon object at 0x7f1e230b7...  
2   6.812127  10.352015  <pysal.cg.shapes.Polygon object at 0x7f1e230b7...  
3  10.084926  12.840340  <pysal.cg.shapes.Polygon object at 0x7f1e230b7...  
4   7.557643  10.313002  <pysal.cg.shapes.Polygon object at 0x7f1e230b7...  

[5 rows x 70 columns]

The `read_files` function only works on shapefile/dbf pairs. If you need to read in data using CSVs, use pandas directly:

In [28]:
usjoin = pd.read_csv(csv_path)
#usjoin = ps.pdio.read_files(csv_path) #will not work, not a shp/dbf pair

In [29]:
usjoin.head()

Name  STATE_FIPS  1929  1930  1931  1932  1933  1934  1935  1936  \
0     Alabama           1   323   267   224   162   166   211   217   251   
1     Arizona           4   600   520   429   321   308   362   416   462   
2    Arkansas           5   310   228   215   157   157   187   207   247   
3  California           6   991   887   749   580   546   603   660   771   
4    Colorado           8   634   578   471   354   353   368   444   542   

   ...     2000   2001   2002   2003   2004   2005   2006   2007   2008   2009  
0  ...    23471  24467  25161  26065  27665  29097  30634  31988  32819  32274  
1  ...    25578  26232  26469  27106  28753  30671  32552  33470  33445  32077  
2  ...    22257  23532  23929  25074  26465  27512  29041  31070  31800  31493  
3  ...    32275  32750  32900  33801  35663  37463  40169  41943  42377  40902  
4  ...    32949  34228  33963  34092  35543  37388  39662  41165  41719  40093  

[5 rows x 83 columns]

The nice thing about working with pandas dataframes is that they have very powerful baked-in support for relational-style queries. By this, I mean that it is very easy to find things like:

The number of counties in each state:

In [30]:
data_table.groupby("STATE_NAME").size()

STATE_NAME
Alabama                  67
Arizona                  14
Arkansas                 75
California               58
Colorado                 63
Connecticut               8
Delaware                  3
District of Columbia      1
Florida                  67
Georgia                 159
Idaho                    44
Illinois                102
Indiana                  92
Iowa                     99
Kansas                  105
Kentucky                120
Louisiana                64
Maine                    16
Maryland                 24
Massachusetts            12
Michigan                 83
Minnesota                87
Mississippi              82
Missouri                115
Montana                  55
Nebraska                 93
Nevada                   17
New Hampshire            10
New Jersey               21
New Mexico               32
New York                 58
North Carolina          100
North Dakota             53
Ohio                     88
Oklahoma                 77
Oregon   

Or, to get the rows of the table that are in Arizona, we can use the `query` function of the dataframe:

In [31]:
data_table.query('STATE_NAME == "Arizona"')

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
1707      Navajo    Arizona         04       017  04017       4      17   
1708    Coconino    Arizona         04       005  04005       4       5   
1722      Mohave    Arizona         04       015  04015       4      15   
1726      Apache    Arizona         04       001  04001       4       1   
2002     Yavapai    Arizona         04       025  04025       4      25   
2182        Gila    Arizona         04       007  04007       4       7   
2262    Maricopa    Arizona         04       013  04013       4      13   
2311    Greenlee    Arizona         04       011  04011       4      11   
2326      Graham    Arizona         04       009  04009       4       9   
2353       Pinal    Arizona         04       021  04021       4      21   
2499        Pima    Arizona         04       019  04019       4      19   
2514     Cochise    Arizona         04       003  04003       4       3   
2615  Santa Cruz    Arizona         04       023  04023       4      23   
3080      La Paz    Arizona         04       012  04012       4      12   

      FIPSNO  SOUTH       HR60  \
1707    4017      0   5.263989   
1708    4005      0   3.185449   
1722    4015      0   0.000000   
1726    4001      0  10.951223   
2002    4025      0   3.458771   
2182    4007      0   6.473749   
2262    4013      0   6.179259   
2311    4011      0   2.896284   
2326    4009      0   4.746648   
2353    4021      0  13.828390   
2499    4019      0   5.520841   
2514    4003      0   4.845049   
2615    4023      0   9.252406   
3080    4012      0   5.046682   

                            ...                             BLK90      GI59  \
1707                        ...                          0.905251  0.366863   
1708                        ...                          1.469081  0.301222   
1722                        ...                          0.324075  0.279339   
1726                        ...                          0.162361  0.395913   
2002                        ...                          0.298011  0.289509   
2182                        ...                          0.246171  0.265294   
2262                        ...                          3.499221  0.277828   
2311                        ...                          0.349650  0.177691   
2326                        ...                          1.890487  0.310256   
2353                        ...                          3.134586  0.304294   
2499                        ...                          3.118252  0.268266   
2514                        ...                          5.201590  0.261208   
2615                        ...                          0.326863  0.327130   
3080                        ...                          2.628811  0.271556   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
1707  0.414135  0.401999  0.445299  13.146998  12.1  13.762783  18.033782   
1708  0.377785  0.381655  0.403188   9.475171   8.5  11.181563  15.267643   
1722  0.347150  0.375790  0.374383  11.508554   4.8   7.018268   9.214294   
1726  0.450552  0.431013  0.489132  15.014738  14.6  18.727548  22.933635   
2002  0.378195  0.376313  0.384089   9.930032   8.6   7.516372   9.483521   
2182  0.337519  0.353848  0.386976  10.470261   8.1   9.934237  11.706102   
2262  0.352374  0.366015  0.372756  10.642382   9.8  11.857260  14.404902   
2311  0.257158  0.283518  0.337256   9.806115   6.7   5.295110  10.453284   
2326  0.362926  0.383554  0.408379  11.979335  10.1  11.961367  16.129032   
2353  0.369974  0.361193  0.400130  10.822965   8.8  10.341699  15.304144   
2499  0.367218  0.375039  0.392144  11.381626  10.2  12.689768  16.163178   
2514  0.359500  0.359701  0.399208  10.197573   8.7   9.912732  13.733872   
2615  0.396807  0.393240  0.413795  19.007213  14.7  15.690913  18.272244   
3080  0.364110  0.372662  0.405743   9.216414   8.0   9.296093  12.379134   

                                 

Behind the scenes, this uses a fast vectorized library, `numexpr`, to essentially do the following. 

First, compare each row's `STATE_NAME` column to `'Arizona'` and return `True` if the row matches:

In [32]:
data_table.STATE_NAME == 'Arizona'

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
3055    False
3056    False
3057    False
3058    False
3059    False
3060    False
3061    False
3062    False
3063    False
3064    False
3065    False
3066    False
3067    False
3068    False
3069    False
3070    False
3071    False
3072    False
3073    False
3074    False
3075    False
3076    False
3077    False
3078    False
3079    False
3080     True
3081    False
3082    False
3083    False
3084    False
Name: STATE_NAME, dtype: bool

Then, use that to filter out rows where the condition is true:

In [33]:
data_table[data_table.STATE_NAME == 'Arizona']

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
1707      Navajo    Arizona         04       017  04017       4      17   
1708    Coconino    Arizona         04       005  04005       4       5   
1722      Mohave    Arizona         04       015  04015       4      15   
1726      Apache    Arizona         04       001  04001       4       1   
2002     Yavapai    Arizona         04       025  04025       4      25   
2182        Gila    Arizona         04       007  04007       4       7   
2262    Maricopa    Arizona         04       013  04013       4      13   
2311    Greenlee    Arizona         04       011  04011       4      11   
2326      Graham    Arizona         04       009  04009       4       9   
2353       Pinal    Arizona         04       021  04021       4      21   
2499        Pima    Arizona         04       019  04019       4      19   
2514     Cochise    Arizona         04       003  04003       4       3   
2615  Santa Cruz    Arizona         04       023  04023       4      23   
3080      La Paz    Arizona         04       012  04012       4      12   

      FIPSNO  SOUTH       HR60  \
1707    4017      0   5.263989   
1708    4005      0   3.185449   
1722    4015      0   0.000000   
1726    4001      0  10.951223   
2002    4025      0   3.458771   
2182    4007      0   6.473749   
2262    4013      0   6.179259   
2311    4011      0   2.896284   
2326    4009      0   4.746648   
2353    4021      0  13.828390   
2499    4019      0   5.520841   
2514    4003      0   4.845049   
2615    4023      0   9.252406   
3080    4012      0   5.046682   

                            ...                             BLK90      GI59  \
1707                        ...                          0.905251  0.366863   
1708                        ...                          1.469081  0.301222   
1722                        ...                          0.324075  0.279339   
1726                        ...                          0.162361  0.395913   
2002                        ...                          0.298011  0.289509   
2182                        ...                          0.246171  0.265294   
2262                        ...                          3.499221  0.277828   
2311                        ...                          0.349650  0.177691   
2326                        ...                          1.890487  0.310256   
2353                        ...                          3.134586  0.304294   
2499                        ...                          3.118252  0.268266   
2514                        ...                          5.201590  0.261208   
2615                        ...                          0.326863  0.327130   
3080                        ...                          2.628811  0.271556   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
1707  0.414135  0.401999  0.445299  13.146998  12.1  13.762783  18.033782   
1708  0.377785  0.381655  0.403188   9.475171   8.5  11.181563  15.267643   
1722  0.347150  0.375790  0.374383  11.508554   4.8   7.018268   9.214294   
1726  0.450552  0.431013  0.489132  15.014738  14.6  18.727548  22.933635   
2002  0.378195  0.376313  0.384089   9.930032   8.6   7.516372   9.483521   
2182  0.337519  0.353848  0.386976  10.470261   8.1   9.934237  11.706102   
2262  0.352374  0.366015  0.372756  10.642382   9.8  11.857260  14.404902   
2311  0.257158  0.283518  0.337256   9.806115   6.7   5.295110  10.453284   
2326  0.362926  0.383554  0.408379  11.979335  10.1  11.961367  16.129032   
2353  0.369974  0.361193  0.400130  10.822965   8.8  10.341699  15.304144   
2499  0.367218  0.375039  0.392144  11.381626  10.2  12.689768  16.163178   
2514  0.359500  0.359701  0.399208  10.197573   8.7   9.912732  13.733872   
2615  0.396807  0.393240  0.413795  19.007213  14.7  15.690913  18.272244   
3080  0.364110  0.372662  0.405743   9.216414   8.0   9.296093  12.379134   

                                 

We might need this behind the scenes knowledge when we want to chain together conditions, or when we need to do spatial queries. 

This is because spatial queries are somewhat more complex. Let's say, for example, we want all of the counties in the US to the West of `-121` longitude. We need a way to express that question. Ideally, we want something like:

```
SELECT
        *
FROM
        data_table
WHERE
        x_centroid < -121
```

So, let's refer to an arbitrary polygon in the the dataframe's geometry column as `poly`. The centroid of a PySAL polygon is stored as an `(X,Y)` pair, so the longidtude is the first element of the pair, `poly.centroid[0]`. 

Then, applying this condition to each geometry, we get the same kind of filter we used above to grab only counties in Arizona:

In [34]:
data_table.geometry.apply(lambda x: x.centroid[0] < -121)

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27       True
28      False
29      False
        ...  
3055    False
3056    False
3057    False
3058    False
3059    False
3060    False
3061    False
3062    False
3063    False
3064    False
3065    False
3066    False
3067    False
3068    False
3069    False
3070    False
3071    False
3072    False
3073    False
3074    False
3075    False
3076    False
3077    False
3078    False
3079    False
3080    False
3081    False
3082    False
3083    False
3084    False
Name: geometry, dtype: bool

If we use this as a filter on the table, we can get only the rows that match that condition, just like we did for the `STATE_NAME` query:

In [35]:
data_table[data_table.geometry.apply(lambda x: x.centroid[0] < -121)]

NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
27          Whatcom  Washington         53       073  53073      53      73   
31           Skagit  Washington         53       057  53057      53      57   
44          Clallam  Washington         53       009  53009      53       9   
47        Snohomish  Washington         53       061  53061      53      61   
48           Island  Washington         53       029  53029      53      29   
57        Jefferson  Washington         53       031  53031      53      31   
71           Kitsap  Washington         53       035  53035      53      35   
80             King  Washington         53       033  53033      53      33   
85            Mason  Washington         53       045  53045      53      45   
92     Grays Harbor  Washington         53       027  53027      53      27   
107          Pierce  Washington         53       053  53053      53      53   
116        Thurston  Washington         53       067  53067      53      67   
130         Pacific  Washington         53       049  53049      53      49   
131           Lewis  Washington         53       041  53041      53      41   
169        Skamania  Washington         53       059  53059      53      59   
170         Cowlitz  Washington         53       015  53015      53      15   
171       Wahkiakum  Washington         53       069  53069      53      69   
181         Clatsop      Oregon         41       007  41007      41       7   
183        Columbia      Oregon         41       009  41009      41       9   
193           Clark  Washington         53       011  53011      53      11   
216       Tillamook      Oregon         41       057  41057      41      57   
217      Washington      Oregon         41       067  41067      41      67   
224       Multnomah      Oregon         41       051  41051      41      51   
225      Hood River      Oregon         41       027  41027      41      27   
226           Wasco      Oregon         41       065  41065      41      65   
240       Clackamas      Oregon         41       005  41005      41       5   
242         Yamhill      Oregon         41       071  41071      41      71   
253          Marion      Oregon         41       047  41047      41      47   
269            Polk      Oregon         41       053  41053      41      53   
272         Lincoln      Oregon         41       041  41041      41      41   
...             ...         ...        ...       ...    ...     ...     ...   
493           Curry      Oregon         41       015  41015      41      15   
517       Josephine      Oregon         41       033  41033      41      33   
594        Siskiyou  California         06       093  06093       6      93   
601       Del Norte  California         06       015  06015       6      15   
687        Humboldt  California         06       023  06023       6      23   
709         Trinity  California         06       105  06105       6     105   
735          Shasta  California         06       089  06089       6      89   
907          Tehama  California         06       103  06103       6     103   
974           Butte  California         06       007  06007       6       7   
1005      Mendocino  California         06       045  06045       6      45   
1047          Glenn  California         06       021  06021       6      21   
1090           Yuba  California         06       115  06115       6     115   
1101           Lake  California         06       033  06033       6      33   
1141         Colusa  California         06       011  06011       6      11   
1172         Sutter  California         06       101  06101       6     101   
1262           Yolo  California         06       113  06113       6     113   
1281           Napa  California         06       055  06055       6      55   
1282         Sonoma  California         06       097  06097       6      97   
1301     Sacramento  California         06       067  06067      

This works on any type of spatial query. 

For instance, if we wanted to find all of the counties that are within a threshold distance from an observation's centroid, we can do it in the following way. 

First, specify the observation. Here, we'll use Cook County, IL:

In [36]:
data_table.query('(NAME == "Cook") & (STATE_NAME == "Illinois")')

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  FIPSNO  \
3044  Cook   Illinois         17       031  17031      17      31   17031   

      SOUTH      HR60                        ...                          \
3044      0  7.108893                        ...                           

          BLK90      GI59      GI69      GI79      GI89       FH60  FH70  \
3044  25.800778  0.259469  0.333628  0.380441  0.381362  14.615451  13.5   

           FH80       FH90                                           geometry  
3044  19.773258  22.473766  <pysal.cg.shapes.Polygon object at 0x7f1e224ca...  

[1 rows x 70 columns]

In [37]:
geom = data_table.query('(NAME == "Cook") & (STATE_NAME == "Illinois")').geometry

In [38]:
geom.values[0].centroid

(-87.82107391263027, 41.84346628270174)

In [39]:
cook_county_centroid = geom.values[0].centroid

In [40]:
import scipy.spatial.distance as d
def near_target_point(polygon, target=cook_county_centroid, threshold=1):
    return d.euclidean(polygon.centroid, target) < threshold 

In [41]:
data_table[data_table.geometry.apply(near_target_point)]

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  FIPSNO  \
631       Will   Illinois         17       197  17197      17     197   17197   
633    Kendall   Illinois         17       093  17093      17      93   17093   
634       Lake    Indiana         18       089  18089      18      89   18089   
635     Porter    Indiana         18       127  18127      18     127   18127   
686     Grundy   Illinois         17       063  17063      17      63   17063   
718   Kankakee   Illinois         17       091  17091      17      91   17091   
731     Newton    Indiana         18       111  18111      18     111   18111   
3010    Racine  Wisconsin         55       101  55101      55     101   55101   
3019   Kenosha  Wisconsin         55       059  55059      55      59   55059   
3028   McHenry   Illinois         17       111  17111      17     111   17111   
3030      Lake   Illinois         17       097  17097      17      97   17097   
3044      Cook   Illinois         17       031  17031      17      31   17031   
3045      Kane   Illinois         17       089  17089      17      89   17089   
3046   De Kalb   Illinois         17       037  17037      17      37   17037   
3059   Du Page   Illinois         17       043  17043      17      43   17043   

      SOUTH      HR60                        ...                          \
631       0  2.783330                        ...                           
633       0  0.000000                        ...                           
634       0  6.299491                        ...                           
635       0  1.658953                        ...                           
686       0  0.000000                        ...                           
718       0  1.810355                        ...                           
731       0  2.898047                        ...                           
3010      0  2.351044                        ...                           
3019      0  0.993888                        ...                           
3028      0  1.583343                        ...                           
3030      0  1.135115                        ...                           
3044      0  7.108893                        ...                           
3045      0  0.960403                        ...                           
3046      0  1.289142                        ...                           
3059      0  0.850723                        ...                           

          BLK90      GI59      GI69      GI79      GI89       FH60  FH70  \
631   10.735965  0.240381  0.289511  0.303416  0.315735   9.818182   7.5   
633    0.532819  0.257018  0.267486  0.281763  0.289925   6.638544   4.5   
634   24.535213  0.234316  0.306393  0.341476  0.369924  10.490816  10.5   
635    0.352124  0.236710  0.289391  0.302400  0.325637   7.351241   6.3   
686    0.064941  0.245480  0.292372  0.311057  0.316300   9.877800   5.4   
718   14.959223  0.236170  0.317750  0.350920  0.358303   9.825252   8.7   
731    0.066416  0.301172  0.319702  0.327321  0.338931  10.296097   7.9   
3010   9.711827  0.232056  0.304189  0.323170  0.343930   9.602611   8.7   
3019   4.130877  0.237876  0.300109  0.319694  0.341010   9.385645   8.7   
3028   0.169176  0.257156  0.306415  0.315220  0.302329   8.096341   6.4   
3030   6.733112  0.249866  0.330083  0.362992  0.317872   7.763546   7.0   
3044  25.800778  0.259469  0.333628  0.380441  0.381362  14.615451  13.5   
3045   5.986689  0.234401  0.290364  0.319262  0.326945   9.189593   7.8   
3046   2.654879  0.260709  0.322043  0.325077  0.329307   9.609443   6.0   
3059   1.978083  0.241338  0.272934  0.303349  0.283952   6.010936   5.2   

           FH80       FH90                                           geometry  
631   10.259602  11.559338  <pysal.cg.shapes.Polygon object at 0x7f1e22cb5...  
633    7.263397   8.099082  <pysal.cg.shapes.Polygon object at 0x7f1e22cb5...  
634   16.840574  20.769121  <pysal.cg.s

### Moving in and out of the dataframe

Most things in PySAL will be explicit about what type their input should be. Most of the time, PySAL functions require either lists or arrays. This is why the file-handler methods are the default IO method in PySAL: the rest of the computational tools are built around their datatypes. 

However, it is very easy to get the correct datatype from Pandas using the `values` and `tolist` commands. 

`tolist()` will convert its entries to a list. But, it can only be called on individual columns (called `Series` in `pandas` documentation)

So, to turn the `NAME` column into a list:

In [42]:
data_table.NAME.tolist()

[u'Lake of the Woods',
 u'Ferry',
 u'Stevens',
 u'Okanogan',
 u'Pend Oreille',
 u'Boundary',
 u'Lincoln',
 u'Flathead',
 u'Glacier',
 u'Toole',
 u'Liberty',
 u'Hill',
 u'Sheridan',
 u'Divide',
 u'Burke',
 u'Renville',
 u'Bottineau',
 u'Rolette',
 u'Towner',
 u'Cavalier',
 u'Pembina',
 u'Kittson',
 u'Roseau',
 u'Blaine',
 u'Phillips',
 u'Valley',
 u'Daniels',
 u'Whatcom',
 u'Bonner',
 u'Ward',
 u'Koochiching',
 u'Skagit',
 u'Williams',
 u'McHenry',
 u'St. Louis',
 u'Roosevelt',
 u'Mountrial',
 u'Marshall',
 u'Ramsey',
 u'Walsh',
 u'Beltrami',
 u'Pierce',
 u'Chelan',
 u'Pondera',
 u'Clallam',
 u'Benson',
 u'Chouteau',
 u'Snohomish',
 u'Island',
 u'Sanders',
 u'Lake',
 u'Nelson',
 u'Grand Forks',
 u'Polk',
 u'Pennington',
 u'Douglas',
 u'McKenzie',
 u'Jefferson',
 u'Richland',
 u'Teton',
 u'McCone',
 u'Shoshone',
 u'Spokane',
 u'Lake',
 u'Clearwater',
 u'Kootenai',
 u'Garfield',
 u'Red Lake',
 u'Grant',
 u'Lincoln',
 u'Lewis and Clark',
 u'Kitsap',
 u'Itasca',
 u'Sheridan',
 u'Wells',
 u'

To extract many columns, you must select the columns you want and call their `.values` attribute. 

If we were interested in grabbing all of the `HR` variables in the dataframe, we could first select those column names:

In [43]:
HRs = [col for col in data_table.columns if col.startswith('HR')]

We can use this to focus only on the columns we want:

In [44]:
data_table[HRs]

HR60       HR70       HR80       HR90
0      0.000000   0.000000   8.855827   0.000000
1      0.000000   0.000000  17.208742  15.885624
2      1.863863   1.915158   3.450775   6.462453
3      2.612330   1.288643   3.263814   6.996502
4      0.000000   0.000000   7.770008   7.478033
5      0.000000   0.000000   4.573101   4.000640
6      7.976390   5.536179   5.633168   5.720497
7      1.011173   1.689475   4.490115   2.814460
8     11.529039   9.273857  28.227324   5.500096
9      0.000000   5.708740   0.000000   6.605892
10     0.000000   0.000000   0.000000   0.000000
11     3.574045   3.840688   7.413585   1.888146
12     0.000000   0.000000   0.000000   0.000000
13     0.000000   0.000000   0.000000   0.000000
14     0.000000   0.000000   0.000000   0.000000
15     0.000000   0.000000   0.000000   0.000000
16     2.945942   0.000000   0.000000   0.000000
17     0.000000   0.000000  19.161808   5.219752
18     0.000000   0.000000   0.000000   0.000000
19     0.000000   8.117213   0.000000   0.000000
20     0.000000   0.000000   0.000000   0.000000
21     0.000000   4.864050   0.000000   0.000000
22     0.000000   0.000000   0.000000   2.218377
23     4.119804   9.910312  14.287755  14.863258
24     5.530668   0.000000  12.421589   0.000000
25     5.854801   5.811757   6.504065   4.045798
26     0.000000  10.811980   0.000000   0.000000
27     1.422131   2.439530   4.061193   2.086920
28     2.138534   2.142245   5.518079   3.756292
29     0.708135   1.138434   0.570854   1.150993
...         ...        ...        ...        ...
3055   0.000000   0.000000   2.107526   0.739919
3056   0.000000   0.970572   4.400324   0.825491
3057   0.611430   2.568301   1.974919   2.828994
3058   2.022286   2.033140   0.000000   3.987956
3059   0.850723   1.981012   2.782690   2.260130
3060   1.790825   3.732470   5.757329   4.007173
3061   0.556604   1.060271   1.010560   2.769193
3062   0.000000   1.756836   5.505395   2.907653
3063   0.427592   3.688132   2.625587   0.000000
3064   0.896660   2.704530   4.229303   2.938915
3065   1.051403   5.379304   7.057466   3.424679
3066   2.095434   6.671377   9.243279   7.285916
3067   1.872835   3.951663   5.339935   3.201065
3068   1.917913   6.382639   1.304631   0.000000
3069   1.939789   4.960564   0.000000   6.072530
3070   5.452765  15.156192  24.841012  28.268787
3071   7.520089   9.163383  10.590286   6.443839
3072   7.202448   9.746302  11.850014  12.561604
3073   8.253379  15.655752  21.173432  16.479507
3074   2.181802   3.074760   3.191133   3.300700
3075   4.902862  11.782264   7.680787  18.362582
3076  18.513376  17.133324  15.034136  12.027015
3077   4.159907   4.126434   3.967782   6.585273
3078   5.403098   5.970974   4.127839   2.586787
3079   1.121183   1.096311   2.442074   2.806112
3080   5.046682  13.152054  13.251761   5.521552
3081   3.411368   7.393533  11.453817   8.691999
3082   1.544425   6.023552   5.280349   4.367330
3083   9.302820   1.800148   3.000030   3.727712
3084   3.396162   2.284879   1.194743   2.048855

[3085 rows x 4 columns]

With this, calling `.values` gives an array containing all of the entries in this subset of the table:

In [45]:
data_table[['HR90', 'HR80']].values

array([[  0.        ,   8.85582713],
       [ 15.88562351,  17.20874204],
       [  6.46245315,   3.4507747 ],
       ..., 
       [  4.36732988,   5.2803488 ],
       [  3.72771194,   3.00003   ],
       [  2.04885495,   1.19474313]])

Using the PySAL pdio tools means that if you're comfortable with working in Pandas, you can continue to do so. 

If you're more comfortable using Numpy or raw Python to do your data processing, PySAL's IO tools naturally support this. 